graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/PaginationResultES7.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.storage.elasticsearch7;
+
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.assistedinject.AssistedInject;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.action.search.SearchRequest;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.action.search.SearchResponse;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.search.SearchHit;
+import org.graylog.shaded.elasticsearch7.org.elasticsearch.search.builder.SearchSourceBuilder;
+
+import java.io.IOException;
+import java.util.List;
+
+public class PaginationResultES7 extends ChunkedQueryResultES7 {
+    private SearchRequest initialSearchRequest;
+
+    @AssistedInject
+    public PaginationResultES7(ElasticsearchClient client,
+                               SearchRequest initialSearchRequest,
+                               @Assisted SearchResponse initialResult,
+                               @Assisted("query") String query,
+                               @Assisted List<String> fields,
+                               @Assisted int limit) {
+        super(client, initialResult, query, fields, limit);
+        this.initialSearchRequest = initialSearchRequest;
+    }
+
+    @Override
+    protected SearchResponse nextSearchResult() throws IOException {
+        final SearchSourceBuilder initialQuery = initialSearchRequest.source();
+        final SearchHit[] hits = lastSearchResponse.getHits().getHits();
+        initialQuery.searchAfter(hits[hits.length - 1].getSortValues());
```

Comment:
Can't we get out of range here?

Reply:
Because of `if (resultMessages.size() == 0) ` check in line 73 of ChunkedQueryResult class it does not cause any problems, but I have introduced additional check anyway, to be on the safe side.

Thanks once again!

Reply:
Awesome, thanks!